server/src/main/java/org/opensearch/cluster/service/Throttler.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.service;
+
+import org.opensearch.common.AdjustableSemaphore;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Base class for Throttling logic.
+ * It provides throttling functionality over multiple keys.
+ *
+ * @param <T> the type of key on which we want to do throttling.
+ */
+public class Throttler<T> {
+    protected ConcurrentMap<T, AdjustableSemaphore> semaphores = new ConcurrentHashMap<T, AdjustableSemaphore>();
+
+    /**
+     * Method to acquire permits for a key type.
+     * It will return true if permits can be acquired within threshold limits.
+     *
+     * If threshold is not set for key then also it will return True.
+     *
+     * @param type Key for which we want to acquire permits.
+     * @param permits Number of permits to acquire.
+     * @return boolean representing was it able to acquire the permits or not.
+     */
+    public boolean acquire(final T type, final int permits) {
+        assert permits > 0;
+        AdjustableSemaphore semaphore = semaphores.get(type);
+        if (Objects.nonNull(semaphore)) {
+            return semaphore.tryAcquire(permits);
+        }
+        return true;
```

Comment:
The default is certainly wrong: the caller should not assume the permit is acquired when there is no semaphore backing it.

Reply:
The problem here is that any mistake in configuration (fe missed key) would go unnoticed. I strongly believe that this class should either return `false` or throw an exception when operates on unknown keys. Moreover, since the key could be (un)intentionally removed, something that was throttled before becomes unthrottled after - silently.

Reply:
> If we return `false` then also issue remain the same, as if key is removed then all the calls will be throttled silently, and can't differentiate as if limit got breached or throttling due to default behavior.

This is not the same semantically - the caller knows nothing will be throttled.

> I am leaning towards taking default behavior in argument. With throwing exception caller will need to handle it every time Throttler is being called, if it is missed some place it can bubble up to the higher layer and cause other issue.

The method could return `Optional<Boolean>` to emphasize that key could be absent and it is up to the caller to decide what the default is.

Reply:
Agree, we can return Optional and let caller decide about the default as caller might need different default behavior for different key. Will make this change. 

Thanks @reta for your inputs in making it right.